precise/iaf_psc_exp_ps_time_reversal.h

```diff
@@ -0,0 +1,589 @@
+    
+    void get_status(DictionaryDatum &) const;
+    void set_status(const DictionaryDatum &);
+    
+  private:
+
+    /** @name Interface functions
+     * @note These functions are private, so that they can be accessed
+     * only through a Node*. 
+     */
+    //@{
+    void init_node_(const Node & proto);
+    void init_state_(const Node & proto);
+    void init_buffers_();
+    void calibrate();
+
+    /**
+     * Time Evolution Operator.
+     *
+     * update() promotes the state of the neuron from origin+from to origin+to.
+     * It does so in steps of the resolution h.  Within each step, time is
+     * advanced from event to event, as retrieved from the spike queue.  
+     *
+     * Return from refractoriness is handled as a special event in the
+     * queue, which is marked by a weight that is GSL_NAN.  This greatly simplifies
+     * the code.
+     *
+     * For steps, during which no events occur, the precomputed propagator matrix
+     * is used.  For other steps, the propagator matrix is computed as needed.
+     *
+     * While the neuron is refractory, membrane potential (y2_) is
+     * clamped to U_reset_.
+     */
+    void update(Time const & origin, const long from, const long to);
+    //@}
+    
+    // The next two classes need to be friends to access the State_ class/member
+    friend class RecordablesMap<iaf_psc_exp_ps_time_reversal>;
+    friend class UniversalDataLogger<iaf_psc_exp_ps_time_reversal>;
+    
+    void set_spiketime(Time const &);
+    
+    /**
+     * Propagate neuron state.
+     * Propagate the neuron's state by dt.
+     * @param dt Interval over which to propagate
+     */
+    void propagate_(const double_t dt);
+    
+    /**
+     * Emit a single spike caused by DC current in absence of spike input.
+     * Emits a single spike and reset neuron given that the membrane
+     * potential was below threshold at the beginning of a mini-timestep
+     * and above afterwards.
+     *
+     * @param origin  Time stamp at beginning of slice
+     * @param lag     Time step within slice
+     * @param t0      Beginning of mini-timestep
+     * @param dt      Duration of mini-timestep
+     */
+    void emit_spike_(const Time & origin, const long lag, 
+		     const double_t t0, const double_t dt);
+    
+    /**
+     * Emit a single spike at a precisely given time.
+     *
+     * @param origin        Time stamp at beginning of slice
+     * @param lag           Time step within slice
+     * @param spike_offset  Time offset for spike
+     */
+    void emit_instant_spike_(const Time & origin, const long lag, 
+			     const double_t spike_offset);
+    
+    /**
+     * Localize threshold crossing by bisectioning.
+     * @param   double_t length of interval since previous event
+     * @returns time from previous event to threshold crossing
+     */
+    double_t bisectioning_(const double_t dt) const;
+
+    void spike_test_count_(const double_t);
+    void spike_test_(const double_t);
```

Comment:
If the function `is_spike_` returns true, spike is emitted by the `emit_spike_` function. `emit_spike_` function uses bisectioning method to determine the time to emit spike [check line 526].